SDL/Event.cu

```diff
@@ -1702,507 +1465,427 @@ unsigned int SDL::Event::getNumberOfQuintupletsByLayerEndcap(unsigned int layer)
+        auto nTrackLocal_buf = allocBufWrapper<int>(devHost, 1, queue);
+        alpaka::memcpy(queue, nTrackLocal_buf, trackCandidatesBuffers->nTrackCandidates_buf, 1);
+        alpaka::wait(queue);
+
+        int nTrackLocal = *alpaka::getPtrNative(nTrackLocal_buf);
+        trackCandidatesInCPU = new SDL::trackCandidatesBuffer<alpaka::DevCpu>(N_MAX_TRACK_CANDIDATES + N_MAX_PIXEL_TRACK_CANDIDATES, devHost, queue);
+        trackCandidatesInCPU->setData(*trackCandidatesInCPU);
+
+        *alpaka::getPtrNative(trackCandidatesInCPU->nTrackCandidates_buf) = nTrackLocal;
+        alpaka::memcpy(queue, trackCandidatesInCPU->hitIndices_buf, trackCandidatesBuffers->hitIndices_buf, 14 * nTrackLocal);
+        alpaka::memcpy(queue, trackCandidatesInCPU->pixelSeedIndex_buf, trackCandidatesBuffers->pixelSeedIndex_buf, nTrackLocal);
+        alpaka::memcpy(queue, trackCandidatesInCPU->logicalLayers_buf, trackCandidatesBuffers->logicalLayers_buf, 7 * nTrackLocal);
+        alpaka::memcpy(queue, trackCandidatesInCPU->directObjectIndices_buf, trackCandidatesBuffers->directObjectIndices_buf, nTrackLocal);
+        alpaka::memcpy(queue, trackCandidatesInCPU->objectIndices_buf, trackCandidatesBuffers->objectIndices_buf, 2 * nTrackLocal);
+        alpaka::memcpy(queue, trackCandidatesInCPU->trackCandidateType_buf, trackCandidatesBuffers->trackCandidateType_buf, nTrackLocal);
+        alpaka::wait(queue);
     }
     return trackCandidatesInCPU;
 }
 
-SDL::trackCandidates* SDL::Event::getTrackCandidatesInCMSSW()
+SDL::trackCandidatesBuffer<alpaka::DevCpu>* SDL::Event::getTrackCandidatesInCMSSW()
 {
     if(trackCandidatesInCPU == nullptr)
     {
-        trackCandidatesInCPU = new SDL::trackCandidates;
-        trackCandidatesInCPU->nTrackCandidates = new unsigned int;
-        cudaMemcpyAsync(trackCandidatesInCPU->nTrackCandidates, trackCandidatesInGPU->nTrackCandidates, sizeof(unsigned int), cudaMemcpyDeviceToHost,stream);
-        cudaStreamSynchronize(stream);
-        unsigned int nTrackCandidates = *(trackCandidatesInCPU->nTrackCandidates);
-
-        trackCandidatesInCPU->trackCandidateType = new short[nTrackCandidates];
-        trackCandidatesInCPU->hitIndices = new unsigned int[14 * nTrackCandidates];
-        trackCandidatesInCPU->pixelSeedIndex = new int[nTrackCandidates];
-
-        cudaMemcpyAsync(trackCandidatesInCPU->hitIndices, trackCandidatesInGPU->hitIndices, 14 * nTrackCandidates * sizeof(unsigned int), cudaMemcpyDeviceToHost, stream);
-        cudaMemcpyAsync(trackCandidatesInCPU->pixelSeedIndex, trackCandidatesInGPU->pixelSeedIndex, nTrackCandidates * sizeof(int), cudaMemcpyDeviceToHost, stream);
-        cudaMemcpyAsync(trackCandidatesInCPU->trackCandidateType, trackCandidatesInGPU->trackCandidateType, nTrackCandidates * sizeof(short), cudaMemcpyDeviceToHost,stream);
-        cudaStreamSynchronize(stream);
+        // Get nTrackLocal parameter to initialize host based trackCandidatesInCPU
```

Comment:
Just local to the function is all. Do you think I should rename or change them to avoid confusion?

Reply:
I guess it confused me, especially since we don't write things in this way for other similar functions. I would rename to omit "Local" from the names, if you agree.

Reply:
Fixed, thank you.